src/main/scala/vsys/blockchain/contract/DataEntry.scala

```diff
@@ -1,74 +1,47 @@
 package vsys.blockchain.contract
 
-import com.google.common.primitives.{Bytes, Ints, Longs, Shorts}
-import play.api.libs.json.{JsObject, JsValue, Json}
+import com.google.common.primitives.{Shorts, Bytes}
+import play.api.libs.json.{JsObject, Json}
 import scorex.crypto.encode.Base58
-import vsys.account.{Address, AddressScheme, ContractAccount, PublicKeyAccount}
-import vsys.account.ContractAccount.ChecksumLength
-import vsys.blockchain.state.ByteStr
 import vsys.blockchain.transaction.ValidationError
 import vsys.blockchain.transaction.ValidationError.InvalidDataEntry
-import vsys.utils.crypto.hash.SecureCryptographicHash._
 
 import scala.util.Success
 
 case class DataEntry(data: Array[Byte],
-                     dataType: DataType.DataTypeVal) {
+                     dataType: DataType.DataTypeVal[_]) {
 
   lazy val bytes: Array[Byte] = Array(dataType.id.asInstanceOf[Byte]) ++ data
 
   lazy val json: JsObject = Json.obj(
-    "data" -> toJson(data, dataType),
+    "data" -> dataType.jsonifierB(data),
     "type" -> dataType
   )
 
-  private def toJson(d: Array[Byte], t: DataType.DataTypeVal): JsValue = {
-    t match {
-      case DataType.PublicKey => Json.toJson(PublicKeyAccount(d).address)
-      case DataType.Address => Json.toJson(Address.fromBytes(d).right.get.address)
-      case DataType.Amount => Json.toJson(Longs.fromByteArray(d))
-      case DataType.Int32 => Json.toJson(Ints.fromByteArray(d))
-      case DataType.ShortText => Json.toJson(Base58.encode(d))
-      case DataType.ContractAccount => Json.toJson(ContractAccount.fromBytes(d).right.get.address)
-      case DataType.TokenId => Json.toJson(ByteStr(d).base58)
-      case DataType.Timestamp => Json.toJson(Longs.fromByteArray(d))
-      case DataType.Boolean => Json.toJson(if (d(0) == 1.toByte) "True" else "False")
-      case DataType.ShortBytes => Json.toJson(Base58.encode(d))
-      case DataType.OpcBlock => Json.toJson("Internal code block")
-      case _ => Json.toJson("Unsupported data type")
-    }
-  }
-
   override def equals(obj: Any): Boolean = obj match {
-    case a: DataEntry => bytes sameElements a.bytes
+    case a: DataEntry => (bytes sameElements a.bytes) && DataType.check(dataType.id.toByte, a.dataType.id.toByte)
```

Comment:
`bytes sameElements a.bytes` is already included `DataType.check(dataType.id.toByte, a.dataType.id.toByte)`?

Reply:
`DataType` byte is not included in bytes of `DataEntry`.

Reply:
check line 14, it should be, right?

Reply:
oh, sorry. i got confused between data and bytes